libs/vkd3d/command.c

```diff
@@ -1139,27 +1140,79 @@ static HRESULT d3d12_command_allocator_allocate_command_buffer(struct d3d12_comm
     return S_OK;
 }
 
-static void d3d12_command_allocator_free_command_buffer(struct d3d12_command_allocator *allocator,
+static HRESULT d3d12_command_allocator_allocate_init_command_buffer(struct d3d12_command_allocator *allocator,
         struct d3d12_command_list *list)
 {
     struct d3d12_device *device = allocator->device;
     const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    VkCommandBufferAllocateInfo command_buffer_info;
+    VkCommandBufferBeginInfo begin_info;
+    VkResult vr;
 
     TRACE("allocator %p, list %p.\n", allocator, list);
 
-    if (allocator->current_command_list == list)
-        allocator->current_command_list = NULL;
+    if (list->vk_init_commands)
+        return S_OK;
+
+    command_buffer_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
+    command_buffer_info.pNext = NULL;
+    command_buffer_info.commandPool = allocator->vk_command_pool;
+    command_buffer_info.level = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+    command_buffer_info.commandBufferCount = 1;
+
+    if ((vr = VK_CALL(vkAllocateCommandBuffers(device->vk_device, &command_buffer_info,
```

Comment:
Can't we just recycle a VkCommandBuffer from the allocator just like normal command lists?

Reply:
Unless I misunderstand what you mean, we do not have a code path for that. `d3d12_command_allocator_allocate_command_buffer` also unconditionally calls `vkAllocateCommandBuffers`.

Reply:
To be clear, there's no recycling going on at all. `d3d12_command_allocator_Reset` calls `vkFreeCommandBuffers`.

Reply:
Ah, I was confused since there is allocator->command_buffers[] which is appended to. I see now it's purely there to vkFreeCommandBuffers on pool reset, which is a little silly. That is orthogonal to this PR, but seems useful to recycle command buffer handles if we can later at some point ...